oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultClientCredentialsTokenResponseClient.java

```diff
@@ -151,120 +97,37 @@ public OAuth2AccessTokenResponse getTokenResponse(OAuth2ClientCredentialsGrantRe
-
-		return new OAuth2Error(errorCode, errorDescription, errorUri);
-	}
-
-	private OAuth2AccessTokenResponse parseTokenResponse(Map<String, String> responseParameters) {
-		if (CollectionUtils.isEmpty(responseParameters) ||
-				!responseParameters.containsKey(OAuth2ParameterNames.ACCESS_TOKEN) ||
-				!responseParameters.containsKey(OAuth2ParameterNames.TOKEN_TYPE)) {
-			return null;
-		}
-
-		String accessToken = responseParameters.get(OAuth2ParameterNames.ACCESS_TOKEN);
-
-		OAuth2AccessToken.TokenType accessTokenType = null;
-		if (OAuth2AccessToken.TokenType.BEARER.getValue().equalsIgnoreCase(
-				responseParameters.get(OAuth2ParameterNames.TOKEN_TYPE))) {
-			accessTokenType = OAuth2AccessToken.TokenType.BEARER;
-		}
-
-		long expiresIn = 0;
-		if (responseParameters.containsKey(OAuth2ParameterNames.EXPIRES_IN)) {
-			try {
-				expiresIn = Long.valueOf(responseParameters.get(OAuth2ParameterNames.EXPIRES_IN));
-			} catch (NumberFormatException ex) { }
-		}
-
-		Set<String> scopes = Collections.emptySet();
-		if (responseParameters.containsKey(OAuth2ParameterNames.SCOPE)) {
-			String scope = responseParameters.get(OAuth2ParameterNames.SCOPE);
-			scopes = Arrays.stream(StringUtils.delimitedListToStringArray(scope, " ")).collect(Collectors.toSet());
-		}
-
-		Map<String, Object> additionalParameters = new LinkedHashMap<>();
-		Set<String> tokenResponseParameterNames = Stream.of(TOKEN_RESPONSE_PARAMETER_NAMES).collect(Collectors.toSet());
-		responseParameters.entrySet().stream()
-				.filter(e -> !tokenResponseParameterNames.contains(e.getKey()))
-				.forEach(e -> additionalParameters.put(e.getKey(), e.getValue()));
-
-		return OAuth2AccessTokenResponse.withToken(accessToken)
-				.tokenType(accessTokenType)
-				.expiresIn(expiresIn)
-				.scopes(scopes)
-				.additionalParameters(additionalParameters)
-				.build();
+	/**
+	 * Sets the {@link Converter} used for converting the {@link OAuth2ClientCredentialsGrantRequest}
+	 * to a {@link RequestEntity} representation of the OAuth 2.0 Access Token Request.
+	 *
+	 * @param requestEntityConverter the {@link Converter} used for converting to a {@link RequestEntity} representation of the Access Token Request
+	 */
+	public final void setRequestEntityConverter(Converter<OAuth2ClientCredentialsGrantRequest, RequestEntity<?>> requestEntityConverter) {
+		Assert.notNull(requestEntityConverter, "requestEntityConverter cannot be null");
+		this.requestEntityConverter = requestEntityConverter;
 	}
 
 	/**
-	 * Sets the {@link RestOperations} used when requesting the access token response.
+	 * Sets the {@link RestOperations} used when requesting the OAuth 2.0 Access Token Response.
 	 *
-	 * @param restOperations the {@link RestOperations} used when requesting the access token response
+	 * @param restOperations the {@link RestOperations} used when requesting the Access Token Response
 	 */
 	public final void setRestOperations(RestOperations restOperations) {
```

Comment:
How can someone inject the RestOperations and still get valid error handling? I think we may need to make OAuth2ErrorResponseErrorHandler public.

Reply:
We should document what the RestOperations required (i.e. the types of converters it would need, error handling, etc)

Reply:
I was thinking of exposing `OAuth2ErrorResponseErrorHandler` but the implementation was pretty easy so decided to hide it.

Either way, I'm fine with exposing it.